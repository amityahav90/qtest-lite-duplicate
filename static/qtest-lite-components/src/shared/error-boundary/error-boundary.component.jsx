import React from 'react';
import PropTypes from 'prop-types';
import { CustomUiLogService } from '../log-service';

export class ErrorBoundary extends React.Component {
	constructor(props) {
		super(props);
		this.state = { error: false };
		this.log = new CustomUiLogService(props.logPath);
	}

	static getDerivedStateFromError() {
		return { error: true };
	}

	componentDidCatch(error, info) {
		this.log.error(error.message, info.componentStack);
		this.setState({ error: true });
	}

	reset() {
		this.setState({ error: false });
	}

	render() {
		const { error } = this.state;
		const { fallbackRender } = this.props;

		if (error) {
			if (fallbackRender) {
				return fallbackRender(this.props, () => {
					this.reset();
				});
			}
		}

		return this.props.children;
	}
}

ErrorBoundary.propTypes = {
	children: PropTypes.any.isRequired,
	logPath: PropTypes.string.isRequired,
	fallbackRender: PropTypes.func
};
