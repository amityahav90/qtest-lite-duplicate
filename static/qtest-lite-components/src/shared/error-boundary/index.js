export * from './error-boundary.component';
export * from './with-error-boundary';
