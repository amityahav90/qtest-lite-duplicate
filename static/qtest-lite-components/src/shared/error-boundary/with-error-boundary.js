import React from 'react';
import { LogContext } from '../log-service';
import { ErrorBoundary } from './error-boundary.component';

export function withErrorBoundary(WrappedComponent, fallbackRender) {
	function wrapped(props) {
		return (
			<LogContext.Consumer>
				{(value) => (
					<ErrorBoundary logPath={value} fallbackRender={fallbackRender}>
						<WrappedComponent {...props} />
					</ErrorBoundary>
				)}
			</LogContext.Consumer>
		);
	}

	return wrapped;
}
