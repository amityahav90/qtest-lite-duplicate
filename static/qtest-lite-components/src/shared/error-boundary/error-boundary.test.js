import React from 'react';
import renderer from 'react-test-renderer';
import { ErrorBoundary } from './error-boundary.component';

describe('Test ErrorBoundary component', () => {
	beforeAll(() => {
		jest.mock('@forge/bridge');
	});

	test('Component rendered correctly', async () => {
		const component = renderer.create(
			<ErrorBoundary logPath="test">
				<div>test</div>
			</ErrorBoundary>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});

	test('"fallbackRender" functionality rendered correctly', async () => {
		function ThrowError() {
			throw new Error('fallbackRender functionality');
		}

		function FallbackRender() {
			return <div>fallbackRender functionality worked</div>;
		}

		const component = renderer.create(
			<ErrorBoundary logPath="test" fallbackRender={() => <FallbackRender />}>
				<ThrowError />
			</ErrorBoundary>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});
});
