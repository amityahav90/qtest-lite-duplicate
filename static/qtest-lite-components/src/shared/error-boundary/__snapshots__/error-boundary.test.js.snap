// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ErrorBoundary component "fallbackRender" functionality rendered correctly 1`] = `
<div>
  fallbackRender functionality worked
</div>
`;

exports[`Test ErrorBoundary component Component rendered correctly 1`] = `
<div>
  test
</div>
`;
